rust_icu_ecma402/src/listformat.rs

```diff
@@ -0,0 +1,234 @@
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+//! Implements the traits found in [ecma402_traits::listformat].
+
+use ecma402_traits;
+use ecma402_traits::listformat;
+use rust_icu_common as common;
+use rust_icu_ulistformatter as ulfmt;
+use std::fmt;
+
+/// Implements [listformat::Format] using ICU as the underlying implementing library.
+#[derive(Debug)]
+pub struct Format {
+    rep: ulfmt::UListFormatter,
+}
+
+// Full support for styled formatting is available since v67.
+#[cfg(feature = "icu_version_67_plus")]
+pub(crate) mod internal {
+    use rust_icu_sys as usys;
+    use ecma402_traits::listformat::options;
+
+    // Converts the param types from ECMA-402 into ICU.
+    pub fn to_icu_width(style: &options::Style) -> usys::UListFormatterWidth {
+        use options::Style;
+        use usys::UListFormatterWidth;
+        match style {
+            Style::Long => UListFormatterWidth::ULISTFMT_WIDTH_WIDE,
+            Style::Short => UListFormatterWidth::ULISTFMT_WIDTH_SHORT,
+            Style::Narrow => UListFormatterWidth::ULISTFMT_WIDTH_NARROW,
+        }
+    }
+
+    // Converts the Type param from ECMA-402 into ICU.
+    pub fn to_icu_type(in_type: &options::Type) -> usys::UListFormatterType {
+        use options::Type;
+        use usys::UListFormatterType;
+        match in_type {
+            Type::Conjunction => UListFormatterType::ULISTFMT_TYPE_AND,
+            Type::Disjunction => UListFormatterType::ULISTFMT_TYPE_OR,
+        }
+    }
+}
+
+impl listformat::Format for Format {
+    type Error = common::Error;
+
+    /// Creates a new [Format], from a [ecma402_traits::Locale] and [listformat::Options].
+    fn try_new<L: ecma402_traits::Locale>(
+        l: L,
+        _opts: listformat::Options,
+    ) -> Result<Format, Self::Error> {
+        let locale = format!("{}", l);
+
+        #[cfg(feature = "icu_version_67_plus")]
+        {
+            let width = internal::to_icu_width(&_opts.style);
+            let in_type = internal::to_icu_type(&_opts.in_type);
+            let rep = ulfmt::UListFormatter::try_new_styled(&locale, in_type, width)?;
+            Ok(Format { rep })
+        }
+
+        // The non-v67 implementation is less featureful.
+        #[cfg(not(feature = "icu_version_67_plus"))]
+        {
+            let rep = ulfmt::UListFormatter::try_new(&locale)?;
+            Ok(Format { rep })
+        }
+    }
+
+    /// Formats the given string.
+    fn format<I, L, W>(self, list: L, f: &mut W) -> fmt::Result
+    where
+        I: fmt::Display,
+        L: IntoIterator<Item = I>,
```

Comment:
The use of `L` for both _locale_ and _list_ in adjacent methods could be confusing.

Reply:
I don't think that is an issue.  Many single-generic parameter functions just us use `T` for the generic type, and that is not confusing.  I think we've been trained so far to pay extra attention to generic types that it should be fine.